pkg/kubevirt/core/util.go

```diff
@@ -138,6 +138,64 @@ ethernets:
 	return interfaces, networks, networkData
 }
 
+const (
+	// regionLabel is the k8s label to specify the region failure domain.
+	regionLabel = "failure-domain.beta.kubernetes.io/region"
+	// zoneLabel is the k8s label to specify the zone failure domain.
+	zoneLabel = "failure-domain.beta.kubernetes.io/zone"
+	// defaultRegion is the name of the default region.
+	// VMs using this region are scheduled on nodes for which a region failure domain is not specified.
+	defaultRegion = "default"
+	// defaultZone is the name of the default zone.
+	// VMs using this zone are scheduled on nodes for which a zone failure domain is not specified.
+	defaultZone = "default"
+)
+
+func buildAffinity(region string, zones []string) *corev1.Affinity {
+	var affinity *corev1.Affinity
+	if region != "" {
+		var matchExpressions []corev1.NodeSelectorRequirement
+		if region != defaultRegion {
+			matchExpressions = append(matchExpressions, corev1.NodeSelectorRequirement{
+				Key:      regionLabel,
+				Operator: corev1.NodeSelectorOpIn,
+				Values:   []string{region},
+			})
+		} else {
+			matchExpressions = append(matchExpressions, corev1.NodeSelectorRequirement{
+				Key:      regionLabel,
+				Operator: corev1.NodeSelectorOpDoesNotExist,
+			})
+		}
+		if len(zones) > 0 {
+			if len(zones) > 1 || zones[0] != defaultZone {
```

Comment:
What is the guarantee that the default zone will be at the index zero?

Reply:
This condition is true if there are multiple zones, or if there is a single zone not named "default". In this case, we target nodes with the zone label within the specified list. Only if there is a single zone and it's named "default", we target nodes without the zone label. If there are multiple zones and one of them is "default" we are still in the first case, which would probably fail (since "default" is an artificial zone), but I decided not to check that here. It could be validated by the validation webhook when creating a shoot.